### PR TITLE
Switch From Clang To GCC

### LINKED ISSUE #1124: Compile Software with gcc instead of clang
### Description of the task
Compile Software with gcc instead of clang

### Acceptance criteria

- [ ] Compiles
- [ ] Coverage works

### PR DESCRIPTION
**Pull request by garethellis0**
### Description
This is basically #1177 with some extra changes to fix test segfaults, build issues, and code coverage.

### Testing Done
Builds and passes in CI, coverage is properly generated.

### Resolved Issues
resolves #1124

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

**Comment by EvanMorcom** [comment]
I think the variable names could be more detailed, but other than that LGTM 💯

File: environment_setup/setup_software.sh
Comment (MathewMacDougall) [inline environment_setup/setup_software.sh:68]: Why is this removed? Didn't we determine this is necessary for `kcachegrind` to work?

File: src/WORKSPACE
Comment (MathewMacDougall) [inline src/WORKSPACE]: Bit of a nitpick. My understanding is that this toolchain is effectively saying "use system gcc". If that's the case should this be named as such? I realize we are making the assumption development takes place on Ubuntu 18 which has gcc 7.4.0 as default, and we install gcc-7 in the setup scripts just in case.

This just might be confusing for people trying to upgrade later?

**Review by jonathanlew — COMMENTED** [review]
Looks a lot cleaner. I'm not sure how you fixed the segfaults but I'm glad you did. Can you also comment on why we need a build switch?

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]

**Comment by MathewMacDougall** [comment]
@jonathanlew segfaults were due to variables in some of my physics tests being destroyed in a different order due to change in compilers. In some low-level tests certain variables must always be destroyed first due to how Box2D manages its pointers.

**Comment by garethellis0** [comment]
> Looks a lot cleaner. I'm not sure how you fixed the segfaults but I'm glad you did. Can you also comment on why we need a build switch?

Sorry, not sure what a "build switch" is?

File: environment_setup/setup_software.sh
Comment (garethellis0) [reply environment_setup/setup_software.sh:68]: Only because the output of `clang` was in a format that required a newer valgrind version. The system one should be fine for anything the system gcc install outputs.

File: src/WORKSPACE
Comment (garethellis0) [reply src/WORKSPACE]: Good point. Changed `linux_gcc_7_4_0` to `linux_gcc` everywhere.

File: .travis.yml
Comment (MathewMacDougall) [inline .travis.yml:15]: just want to double check. Bazel's CLI-Reference says the `coverage` command "Generates code coverage report for specified test targets". It seems it's roughly the same as `test` but with coverage. Does this mean we're not building _all_ the code, ie. missing the top-level code not associated with tests?

I know we need to build code with covarege enabled to get the coverage data (so doesn't make sense to necessarily call `build` here).

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
another question for my sanity

File: .travis.yml
Comment (garethellis0) [reply .travis.yml:15]: Correct, I just thought  of this while I was cleaning this up.......... 

We _could_ run ` bazel build //software:full_system_main` as well, but the different command line options would probably trigger a re-compile of basically everything.

File: src/firmware/main/shared/physics.h
Comment (jonathanlew) [inline src/firmware/main/shared/physics.h]: Can you explain this and why we can't have one version the  or the other?

File: .travis.yml
Comment (MathewMacDougall) [reply .travis.yml:15]: Maybe worth trying locally. Or perhaps we can somehow pass the coverage flags to `build` first? My understanding is that the main difference is the extra symbols / tracking built in for coverage to consume

**Comment by jonathanlew** [comment]
> > Looks a lot cleaner. I'm not sure how you fixed the segfaults but I'm glad you did. Can you also comment on why we need a build switch?
> 
> Sorry, not sure what a "build switch" is?

Commented in line

File: src/firmware/main/shared/physics.h
Comment (garethellis0) [reply src/firmware/main/shared/physics.h]: Ah, we can the one meant for c++ (the `else` case)`, but I think it was still a good option to make the array lengths clear like this when possible, so I left the C version in.

**Review by MathewMacDougall — DISMISSED** [review]

File: src/firmware/main/shared/physics.h
Comment (jonathanlew) [reply src/firmware/main/shared/physics.h]: seems like we can achieve the same clarity with some nice documentation

Comment (garethellis0) [reply src/firmware/main/shared/physics.h]: I don't feel super strongly either way here, summoning @MathewMacDougall and @akhilveeraghanta  for another opinion

Comment (MathewMacDougall) [reply src/firmware/main/shared/physics.h]: I'm personally less concerned about the clarity and more concerned with whether or not the variable length arrays give us any other benefits.

The way we use the VLAs in our implementation doesn't seem to use any of the small additional benefits I'm aware of, such as automatically creating an array whose lifetime is managed in the scope of the function. (been looking at https://stackoverflow.com/questions/16672322/malloced-array-vs-variable-length-array). The only benefit we're getting is that the function signature is _slightly_ more explicit.

So for me it comes down to if I dislike `ifdefs` more than I dislike the lack of extra clarity. Assuming this file will be modified and updates a reasonable amount moving forwards, I dislike `ifdefs` more and therefore would prefer we switch everything to be compatible with both C and C++

Comment (akhilveeraghanta) [reply src/firmware/main/shared/physics.h]: I agree as well, it is in the shared folder so it defeats the purpose being compatible between C and C++ (similar to how constants.h is shared).

If this is specific to c++ then it doesn't belong in the shared folder anyway, so also in favor about removing `ifdefs`

**Review by akhilveeraghanta — DISMISSED** [review]
LGTM :shipit:

File: src/firmware/main/shared/physics.h
Comment (garethellis0) [reply src/firmware/main/shared/physics.h]: Great, seems like there is consensus here, removed the C-specific version!

File: .travis.yml
Comment (garethellis0) [reply .travis.yml:15]: Just made some changes to `full_system_main.cpp` and it failed. No test depends on that target, so looks like bazel builds everything. So this is actually a non-issue! :tada:

**Review by jonathanlew — DISMISSED** [review]
LGTM pending CI :+1:

**Comment by MathewMacDougall** [comment]
Ok so after some investigation it turns out you can't pass matrices as `float**` in c++. This is the cause of the CI failure.

Eg.
```
void matrix_test(int rows, int cols, float** matrix){
    for (int i = 0; i < rows; i++){
        for (int j = 0; j < cols; j++){
            volatile float a = matrix[i][j];
        }
    }
}
int main(int argc, char ** argv) {
    int num_rows = 10;
    int num_cols = 10;
    float matrix[num_rows][num_cols];
    for (int i = 0; i < num_rows; i++) {
        for (int j = 0; j < num_cols; j++){
            matrix[i][j] = i*j;
        }
    }
    matrix_test(num_rows, num_cols, matrix);
}
```
This is valid C code, but invalid C code. (although C does throw a warning).

The easiest solution was to move the offending matrix / linear algebra code to a new file and build target, so it can be purely built as C and use Variable Length Arrays. We should consider proper "Shareable" matrix and vector types moving forwards, but that's out of scope for this PR.

The commits I pushed move the matrix functions to a new file, and also removed any unused functions.

**Review by MathewMacDougall — APPROVED** [review]
:shipit:

**Review by jonathanlew — APPROVED** [review]
:+1:

**Pull request merged**
